keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,435 @@
+through STDOUT.
+
+- Add 2 fields to `kustomization.yaml`: `generators` and `transformers`
+- Plugins to actuate new fields have the path `$XDG_CONFIG_HOME/kustomize/plugins/<API group>`
+
+Much of the existing Kustomize functionality could be implemented through this plugin mechanism.
+
+Example Built-In Generators:
+
+- `resource`
+- `bases`
+- `secretGenerators`
+- `configMapGenerators`
+
+Example Built-In Transforms:
+
+- `commonAnnotations`
+- `commonLabels`
+- `images`
+- `namePrefix` # not supported in iteration 1 due to transformation restrictions
+- `nameSuffix` # not supported in iteration 1 due to transformation restrictions
+- `patches`
+- `namespace` # not supported in iteration 1 due to transformation restrictions
+
+
+## Motivation
+
+1. Enable organizations to develop their own authoring solutions bespoke to their requirements, and to integrate
+   those solutions with kustomize and kubectl.
+1. Enable tools and authoring solutions developed by the ecosystem to natively integrate with kustomize and kubectl
+   commands.
+1. Enable Kustomize power users to augment Kustomize with their own *Transformers* and *Generators*.
+
+### Background
+
+Many users of Kubernetes require the ability to also write Config through high-level abstractions rather
+than only as low-level APIs, or to configure Kubernetes through in-house developed platforms.
+
+Authoring solutions have already been developed:
+
+- In the Ecosystem
+  - Helm
+  - Ksonnet
+- Internally by Organizations
+  - [kube-gen] (AirBnB) 
+
+Support for connecting the tools developed by the Ecosystem with kubectl relies on piping commands together,
+however pipes are an imperative technique that require their own scripting and tooling.
+
+This KEP proposes a declarative approach for connecting tools built in the Kubernetes ecosystem or bespoke
+internal tools developed by organizations.
+
+[kube-gen]: https://events.linuxfoundation.org/wp-content/uploads/2017/12/Services-for-All-How-To-Empower-Engineers-with-Kubernetes-Melanie-Cebula-Airbnb.pdf
+
+### Goals
+
+- Allow Config authoring solutions developed in the ecosystem to be declaratively accessed by kubectl as
+  Generators and Transformers
+- Allow users and organizations to develop their own custom config authoring solutions and integrate
+  them into kubectl
+- Allow Kustomize power users to augment Kustomize's Built-Tranformers
+- Allow users to perform client-side mutations so they they show up in diffs and are auditable
+
+### Non-Goals
+
+- Rewrite Kustomize on top of plugins
+  - This should be possible, but isn't a problem we need to solve right now
+- Develop a market place of any sort
+  - The ecosystem solutions themselves will have market places.  This allows those ecosystem tools to
+    plug their market places into kubectl via Kustomize.
+
+## Proposal
+
+Introduce an executable plugin that has 2 subcommands.
+
+- `$ team.example.com generate`
+- `$ team.example.com transform`
```

Comment:
`team.example.com` is a binary that is suppose to exist `~/kustomize/plugins` ?

Reply:
right